### PR TITLE
Cloudant Geospatial

### PR DESCRIPTION
**Pull request by Danwakeem**
### Description
Added a mixin for Cloudant Geospatial queries.

#### Related issues

- connect to strongloop/loopback#190

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by Danwakeem** [comment]
I wasn't sure what I should do, but I think the errors in the unit tests are from downstream issues with `loopback-connector-couchdb2`. When I only run the tests for this repo and ignore the ones from `loopback-connector-couchdb2` the tests all pass.

**Comment by dhmlau** [comment]
@Danwakeem , FYI - below is the error message of the failed test case:
```
  1) cloudant view.test - imported from couchdb2 couchdb2 view viewDocs returns result by quering a view:
      Uncaught AssertionError: expected 8 to be 4
      + expected - actual
      -8
      +4
      
      at Assertion.fail (node_modules/should/lib/assertion.js:180:17)
      at Assertion.prop.value (node_modules/should/lib/assertion.js:65:17)
      at node_modules/loopback-connector-couchdb2/test/view.test.js:56:37
      at Request._callback (node_modules/cloudant-nano/lib/nano.js:217:16)
      at Request.self.callback (node_modules/request/request.js:185:22)
```

File: lib/geo.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+'use strict';
+
+var URL = require('url');
+var assert = require('assert');
+var util = require('util');
+var _ = require('lodash');
+
+module.exports = mixinGeo;
+
+function mixinGeo(CouchDB) {
```
Comment (dhmlau) [inline lib/geo.js]: If this is meant for cloudant, should this be `Cloudant` instead of `CouchDB`?

Comment (Danwakeem) [reply lib/geo.js]: Yes, you are right. I went ahead and amended the PR.

**Comment by Danwakeem** [comment]
@dhmlau is there anything I need to do for that? I wasn't sure how it should be addressed since it is a test for another module.

**Comment by dhmlau** [comment]
@Danwakeem , thanks for incorporating my comments.  
@jannyHou , could you please review?  Thanks!

**Comment by dhmlau** [comment]
@Danwakeem , I'm not very familiar with this, so I've asked @jannyHou to take a look.  Another thought.. Would it make sense to apply your PR on the couchdb connector `loopback-connector-couchdb2` ?

**Comment by Danwakeem** [comment]
@dhmlau and @jannyHou I found the issue. 

It looks like I was using the same key name in my test objects the `viewDocs` test was expecting to find in its own test, so I was unintentionally adding my data to that test. 

I have added another commit to this PR to address this issue :)

**Comment by Danwakeem** [comment]
Well maybe there is another issue. It looks like the Node 6 build failed for some reason. I’ll keep looking!

**Comment by dhmlau** [comment]
You see the error when you test locally, right?  anyway, the log on Travis didn't say what's wrong, so I kicked off the test again.

**Comment by Danwakeem** [comment]
No, that is what kind has me confused. When I run the tests locally they pass. The error on Travis was a socket hang up so I am not sure what is causing that yet.

**Comment by Danwakeem** [comment]
It looks like I have node 8 on my local machine though. I am going to pull down node 6 and re run locally.

**Comment by Danwakeem** [comment]
It is strange, when I ran the build for my fork in a [new travis pipeline](https://travis-ci.org/Danwakeem/loopback-connector-cloudant) the tests passed. Do you think it could be something with Travis causing the build to fail?

**Comment by virkt25** [comment]
Triggered travis again and it's all green (must've been something on their end). 

@Danwakeem Can you please remove the `package-lock.json` file from this PR. We choose to not use them. If you like you can include a `.npmrc` file with `package-lock=false` to avoid it's generation all together. 

@jannyHou Can you do a review? I looked over it and looks good (with the removal of package-lock.json) overall.

**Comment by Danwakeem** [comment]
@virkt25 Done! I went ahead and added it to the `.npmrc` file as well as the `.gitignore` file.

**Comment by dhmlau** [comment]
@slnode test please

File: test/geo.test.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      db = global.getDataSource();
+      var connector = db.connector;
+
+      db.once('connected', function(err) {
+        async.series([insertSampleData, insertViewDdoc], done);
+      });
+
+      function insertSampleData(cb) {
+        sampleData = generateSamples();
+        connector[connector.name].use(connector.getDbName(connector))
```
Comment (jannyHou) [inline test/geo.test.js]: nitpick: 
`var driverInstance = connector[connector.name].use(connector.getDbName(connector))`

then you can call `driverInstance.bulk`, `driverInstance.insert`

File: lib/geo.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+   * @callback
+   * @param {Function} cb
+   */
+  Cloudant.prototype.geoDocs = function(ddocName, viewName, options, cb) {
+    // omit options, e.g. ds.geoDocs(ddocName, viewName, cb);
+    if (typeof options === 'function' && !cb) {
+      cb = options;
+      options = {};
+    }
+    debug('Cloudant.prototype.view ddocName %s viewName %s options %s',
```
Comment (jannyHou) [inline lib/geo.js]: `Cloudant.prototype.geoDocs`

**Comment by Danwakeem** [comment]
Hey @jannyHou, I have made the adjustments you had suggested above. 

I also went ahead and cleaned up the comments for the `geoDocs` function. I used the `view.js` file as a starting place and didn't realize I left some things in there.

Let me know what you guys think about this round of changes. :)

**Comment by dhmlau** [comment]
I was surprised that CI all green previously.  I just submitted a PR to update the node versions we use in Travis: https://github.com/strongloop/loopback-connector-cloudant/pull/192.  Once that land, we can run the tests in this PR again.

**Comment by Danwakeem** [comment]
@dhmlau would you want me to pull in the `update-travis` branch or just wait for that PR to get merged into master?

**Comment by dhmlau** [comment]
@Danwakeem , the PR #192 just landed. Do you mind to rebase?

**Comment by dhmlau** [comment]
@Danwakeem , the PR #192 just landed. Do you mind to rebase?

**Comment by Danwakeem** [comment]
This may be a silly question but I am not sure how to rebase this. 

I tried running the following commands
```
> git fetch upstream
> git rebase origin master
> git push origin master
```
I must have missed a step there because the PR Linter still says that I need a rebase?

**Comment by Danwakeem** [comment]
Whoops, sorry about that I accidentally went too far back when I tried resetting my fork.

**Comment by Danwakeem** [comment]
I guess that is one way to get the rebase to work lol :)

**Comment by Danwakeem** [comment]
Yeah no problem thank you guys for helping me get that merged in! 😎

**Comment by MarianZoll** [comment]
@Danwakeem @jannyHou 

Great to see this feature implemented with `ds.connector` in v2.1.0. Is there any effort done yet to add it to the `README.md`?

**Comment by virkt25** [comment]
@MarianZoll Not that I am aware of. Would you like to open a PR with the changes to the `README.md`?

**Comment by Danwakeem** [comment]
@virkt25 @MarianZoll I went ahead and created another PR for the updates to the README. Since the API is so similar to views I just followed the same format for the `view` section.

**Pull request merged**